tools/build_script_generator/common.py

```diff
@@ -0,0 +1,218 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2018, Niklas Hauser
+#
+# This file is part of the modm project.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+# -----------------------------------------------------------------------------
+
+import os
+from collections import defaultdict
+
+def common_source_files(env, buildlog):
+    files_to_build = []
+
+    prev = env.outbasepath
+    env.outbasepath = "."
+    outpath = env.outpath("modm")
+    for operations in buildlog:
+        filename = os.path.relpath(operations.filename_out, outpath)
+        _, extension = os.path.splitext(filename)
+
+        if extension in [".c", ".cpp", ".cc", ".sx", ".S"]:
+            files_to_build.append(os.path.normpath(filename))
+
+    env.outbasepath = prev
+    return sorted(files_to_build)
+
+def common_target(target):
+    core = target.get_driver("core")["type"]
+    core = core.replace("fd", "").replace("f", "")
+    mcu = target._properties.get("mcu", "")
+    p = {
+        "core": core,
+        "mcu": mcu,
+        "platform": target.identifier["platform"],
+        "family": target.identifier["family"],
+        "partname": target.partname,
+    }
+    return p
+
+def common_memories(target):
+    core_driver = target.get_driver("core")
+    memories = []
+    if "memory" in core_driver:
+        memories.extend([
+            {
+                k:(int(v) if v.isdigit() else (int(v, 16) if v.startswith("0x") else v))
+                for k, v in memory.items()
+            }
+            for memory in core_driver["memory"]
+        ])
+    return memories
+
+
+def common_compiler_flags(compiler, target, buildlog):
```

Comment:
These are the compiler flags for modm, I moved some into the debugging profile and merged the AVR and ARM options. I cross-checked most of these with the documentation, added/removed a few.

Comments? @rleh @strongly-typed

Reply:
@salkinium: Although I'm neither @rleh nor @strongly-typed, I would like to add some comments. If contribution from my side is not desired, feel free too ignore the comments :wink: 

I have just worked through the online documentation of gcc warning options. The first thing I noticed is that some specified flags are already implied by `-Wall -Wextra` (according to gcc docs) and can be removed (at least for gcc versions that can build modm). This could make the list easier to grasp. This applies to:

-  `-Wformat` implied by `-Wall`
-  `-Winit-self` implied by `-Wall`
-  `-Wuninitialized` implied by `-Wall`
-  `-Wunused-parameter` implied by `-Wunused -Wextra`
-  `-Wwrite-strings` is enabled by default 
-  `-Wimplicit` implied by `-Wall`
-  `-Wdelete-non-virtual-dtor` implied by `-Wall`

Are you sure `-fdata-sections` and `-ffunction-sections` are really beneficial? The docs say the following about them:
>Only use these options when there are significant benefits from doing so.  When you specify these options, the assembler and linker create larger object and executable files and are also slower. 

I also found some flags that we could try:
- `-Wduplicated-cond`: Detects duplicate conditions in `if / else if` statements
- `-Wbad-function-cast`: C only, warn on suspicious cast of return values (e.g. int to pointer cast)
- `-Wlogical-op`: Detect suspicious use of logical operators